bindings/rust/s2n-tls/src/config.rs

```diff
@@ -291,6 +298,76 @@ impl Builder {
         Ok(self)
     }
 
+    /// Corresponds to [s2n_config_add_cert_chain_and_key_to_store].
+    pub fn load_chain(&mut self, chain: CertificateChain<'static>) -> Result<&mut Self, Error> {
+        let add_result = unsafe {
+            s2n_config_add_cert_chain_and_key_to_store(
+                self.as_mut_ptr(),
+                // SAFETY: audit of add_to_store shows that the certificate chain
+                // is not mutated. https://github.com/aws/s2n-tls/issues/4140
+                chain.as_ptr() as *mut _,
+            )
+            .into_result()
+        };
+
+        // Out of an abudance of caution, we hold a reference to the CertificateChain
+        // regardless of whether add_to_store fails or succeeds. We have limited
+        // visibility into the failure modes, so this behavior ensures that _if_
+        // the C library held the reference despite the failure, it would continue
+        // to be valid memory.
+        self.context_mut().application_owned_certs.push(chain);
+
+        add_result?;
+
+        Ok(self)
+    }
+
+    /// Corresponds to [s2n_config_set_cert_chain_and_key_defaults].
+    pub fn set_default_chains<T: IntoIterator<Item = CertificateChain<'static>>>(
+        &mut self,
+        chains: T,
+    ) -> Result<&mut Self, Error> {
+        // Must match S2N_CERT_TYPE_COUNT in s2n_certificate.h
+        const CERT_TYPE_COUNT: usize = 3;
+
+        let mut pointer_array = [std::ptr::null_mut(); CERT_TYPE_COUNT];
+        let mut pointer_array_length = 0;
+
+        for chain in chains.into_iter() {
+            if pointer_array_length >= CERT_TYPE_COUNT {
+                // drop the reference counts that we previously had taken
+                for _ in 0..CERT_TYPE_COUNT {
+                    self.context_mut().application_owned_certs.pop();
+                }
+
+                return Err(Error::bindings(
+                    ErrorType::UsageError,
+                    "InvalidInput",
+                    "A single default can be specified for RSA, ECDSA, 
+                    and RSA-PSS auth types, but more than 3 certs were supplied",
+                ));
+            }
+
+            // SAFETY: manual inspection of set_defaults shows that certificates
+            // are not mutated. https://github.com/aws/s2n-tls/issues/4140
+            pointer_array[pointer_array_length] = chain.as_ptr() as *mut _;
+            pointer_array_length += 1;
+
+            self.context_mut().application_owned_certs.push(chain);
```

Comment:
I don't know that I'm sold on this complexity being worth it. Could we maybe simplify a bit? Choose a reasonable limit not required to be bound to CERT_TYPE_COUNT, like just arbitrarily 10, and error if the iterator is larger than that. And why can't we add all the chains to application_owned_certs after we've made that check, so that we don't have to pop them in case of failure?

Reply:
> Choose a reasonable limit not required to be bound to CERT_TYPE_COUNT

Yup! Can do.

> And why can't we add all the chains to application_owned_certs after we've made that check

This specific flow is being done to [avoid any allocations](https://github.com/aws/s2n-tls/pull/4937#discussion_r1868683145). There isn't any nice way to get the length of an iterator, so we'd have to collect the results into an intermediate (stack allocated) variable. 
```

let stack_of_cert_chain = ...;
// build up cert chain, error if length > 10

let stack_of_ptr = ...;

// append stack_of_cert_chain to application_owned_certs
// this has to occur _after_ the pointers have been extracted

// call set_default
```

So we incur an additional intermediate variable. I can go ahead and implement that and we can see what it looks like?

Reply:
I'd say give it a shot :+1: